haskell/test/SMT/Common/Run.hs

```diff
@@ -0,0 +1,263 @@
+        <> pretty ligoGovBalance <> "\n"
+        )
+
+    -- View contract storage could be updated via registry lookup registry call.
+    when (haskellViewStore /= ligoViewStore) $
+      error
+        (  "*** Error: Haskell and Ligo view contract storage are different:\n"
+        <> modelCallMsg
+        <> "* Haskell view contract storage:\n"
+        <> pretty haskellViewStore <> "\n"
+        <> "* Ligo view contract storage:\n"
+        <> pretty ligoViewStore <> "\n"
+        )
+
+    where
+      modelCallMsg = "* Call with:\n" <> (pretty mc)
+
+
+-- | Advance nettest level and call ligo dao with the provided argument.
+-- Return the result of the call (storage or error) and the storage of
+-- auxiliary contracts.
+handleCallViaLigo
+  :: (TAddress Parameter, TAddress FA2.Parameter, TAddress (MText, Maybe MText))
+  -> ModelCall
+  -> EmulatedT PureM (Either ModelError Storage, Mutez, [FA2.TransferParams], Mutez, [Text])
+handleCallViaLigo (dao, gov, viewC) mc = do
+  case (mc & mcAdvanceLevel) of
+    Just lvl -> advanceLevel lvl
+    Nothing -> pure ()
+
+  nettestResult <- attempt @NettestFailure $ callLigoEntrypoint mc dao
+  let result = parseNettestError nettestResult
+  fs <- getFullStorage (unTAddress dao)
+  let fsE = case result of
+        Just err -> Left err
+        Nothing -> Right (fs & fsStorage)
+
+  daoBalance <- getBalance dao
+
+  govStore <- fromVal @[FA2.TransferParams] <$> getStorage' @(ToT [FA2.TransferParams]) gov
+  govBalance <- getBalance gov
+
+  viewStorage <- fromVal @[(MText, Maybe MText)] <$> getStorage' @(ToT [(MText, Maybe MText)]) viewC
+  pure (fsE, daoBalance, govStore, govBalance, show <$> viewStorage)
+
+
+callLigoEntrypoint :: MonadNettest caps base m => ModelCall -> TAddress Parameter -> m ()
+callLigoEntrypoint mc dao = withSender (mc & mcSource & msoSender) $ case mc & mcParameter of
+  XtzAllowed (Propose p) -> call dao (Call @"Propose") p
+  XtzAllowed (Transfer_contract_tokens p) -> call dao (Call @"Transfer_contract_tokens") p
+  XtzAllowed (Transfer_ownership p) -> call dao (Call @"Transfer_ownership") p
+  XtzAllowed (Accept_ownership p) -> call dao (Call @"Accept_ownership") p
+
+  XtzForbidden (Vote p) -> call dao (Call @"Vote") p
+  XtzForbidden (Flush p) -> call dao (Call @"Flush") p
+  XtzForbidden (Freeze p) -> call dao (Call @"Freeze") p
+  XtzForbidden (Unfreeze p) -> call dao (Call @"Unfreeze") p
+  XtzForbidden (Update_delegate p) -> call dao (Call @"Update_delegate") p
+  XtzForbidden (Drop_proposal p) -> call dao (Call @"Drop_proposal") p
+
+  XtzAllowed (CallCustom p) -> call dao (Call @"CallCustom") p
+
+parseNettestError :: Either NettestFailure a -> Maybe ModelError
+parseNettestError = \case
+  Right _ -> Nothing
+  Left (NettestFailedWith _ (MM.ExpressionString tval)) ->
+    Just $ contractErrorToModelError tval
+  Left (NettestFailedWith _ (MM.ExpressionPrim (MM.MichelinePrimAp (MM.MichelinePrimitive "Pair") (MM.ExpressionString tval :<| _) _))) ->
```

Comment:
I was trying to get the error string from `NettestFailure`. I've tried `fromExpression` but couldn't make it work, could you give me an example of this?

Reply:
Ah ok, there's actually a small problem here. 

In this case, because we don't know/care what the pair's right type is, we can't deserialize the expression to a typed `Value t`, only to an untyped `Value`.
However, in the version of morley you're currently using, an expression can only be deserialized to typed `Value t`, not to `Value`.

When we upgrade to the next version of morley, deserialization an expression to an untyped `Value` [will be possible](https://gitlab.com/morley-framework/morley/-/merge_requests/840) (thanks to @DMozhevitin), and something like this should be possible:

```hs
  Left (NettestFailedWith _ expr) ->
    case MM.fromExpression @U.Value expr of
      Right (U.ValueString s) -> undefined
      Right (U.ValuePair (U.ValueString s) _) -> undefined
```

Can you add a comment on the code, tag the comment with the number of the issue/PR @sras is working on?

@sras , would you mind updating this bit when you rebase your PR on master (after this PR is merged)?

Reply:
Added a comment for this.